dace/libraries/onnx/op_implementations/reduction_ops.py

```diff
@@ -0,0 +1,429 @@
+                                     tasklet_code,
+                                     language=dace.Language.CPP)
+
+        # Add edges for axes input, data input and output
+        nstate.add_edge(data_read, None, tasklet, 'inp', nsdfg.make_array_memlet(data_read.data))
+        nstate.add_edge(axes_node, None, tasklet, 'axes_arr', nsdfg.make_array_memlet(axes_node.data))
+        nstate.add_edge(tasklet, 'out', reduced_write, None, nsdfg.make_array_memlet(reduced_write.data))
+
+        return nsdfg
+
+
+@op_implementation(op="ReduceMean", name="pure")
+class PureReduceMean(ONNXForward):
+    '''
+        ReduceMean expansion
+    '''
+
+    @staticmethod
+    def forward_can_be_applied(node: onnx_op.ONNXOp, state: SDFGState, sdfg: SDFG) -> bool:
+        # Check that all the inputs (even the optional ones) are present and constant
+        # optional inputs
+        is_axes_present = True
+        try:
+            if hasattr(sdfg, "_parent_onnx_model") and in_edge_with_name(
+                    node, state, "axes").src.data not in sdfg._parent_onnx_model.clean_weights:
+                return False
+        except ValueError:
+            is_axes_present = False
+
+        if not is_axes_present and hasattr(node, "axes"):
+            is_axes_present = True
+
+        # Current constraints: axes must be explict. Axes must be zero
+        if not is_axes_present:
+            return False
+
+        return True
+
+    @staticmethod
+    def forward(node: onnx_op.ONNXOp, state: SDFGState, sdfg: SDFG) -> typing.Union[Node, SDFG]:
+        # We treat both cases where axes is an attribute and where it is an input
+        # Since can be applied is true, we know that axes is present and valid
+        axes = None
+        # TODO: avoid catching Exceptions
+        try:
+            if hasattr(sdfg, "_parent_onnx_model") and in_edge_with_name(
+                    node, state, "axes").src.data in sdfg._parent_onnx_model.clean_weights:
+                axes = sdfg._parent_onnx_model.clean_weights[in_edge_with_name(node, state, "axes").src.data].numpy()
+        except ValueError:
+            pass
+        if axes is not None:
+            if len(axes) == 1:
+                axes = axes[0]
+            else:
+                raise NotImplementedError(
+                    "PureReduceMean in the case where there are multiple axes as input connectors is not implemented yet."
+                )
+        else:
+            # Axes is an attribute of the node
+            axes = node.axes if hasattr(node, "axes") else None
+
+        def prog(data, reduced):
+            reduced[:] = np.mean(data, axis=axes)
+
+        result = program_for_node(prog, sdfg, state, node)
+        return result
+
+
+# ============================================================================
+# ReduceSum Operations
+# ============================================================================
+
+
+@op_implementation(op="ReduceSum", name="pure")
```

Comment:
annoys me that this is tagged as "pure" when it is definitely not a native SDFG implementation.